packages/turbo/android/src/main/java/com/reactnativeturbowebview/RNSession.kt

```diff
@@ -1,102 +1,156 @@
+      else -> TurboVisitOptions()
+    }
+
+    viewTreeLifecycleOwner?.lifecycleScope?.launch {
+      val snapshot = when (options.action) {
+        TurboVisitAction.ADVANCE -> fetchCachedSnapshot(url)
+        else -> null
+      }
+
+      viewTreeLifecycleOwner.lifecycle.whenStateAtLeast(Lifecycle.State.STARTED) {
+        turboSession.visit(
+          TurboVisit(
+            location = url,
+            destinationIdentifier = url.hashCode(),
+            restoreWithCachedSnapshot = restoreWithCachedSnapshot,
+            reload = reload,
+            callback = this@RNSession,
+            options = options.copy(snapshotHTML = snapshot)
+          )
+        )
+      }
+    }
+  }
+
+  private suspend fun fetchCachedSnapshot(url: String): String? {
+    return withContext(Dispatchers.IO) {
+      val response = turboSession.offlineRequestHandler?.getCachedSnapshot(
+        url = url
+      )
+      response?.data?.use {
+        String(it.readBytes())
+      }
+    }
+  }
+
+  fun restoreCurrentVisit(): Boolean {
+    return turboSession.restoreCurrentVisit(callback = this)
+  }
+
   inner class JavaScriptInterface {
     @JavascriptInterface
     fun postMessage(messageStr: String) {
       // Android interface works only with primitive types, that's why we need to use JSON
       val messageObj =
-        Utils.convertJsonToBundle(JSONObject(messageStr)) // TODO remove double conversion
-      sendMessage("sessionMessage${sessionHandle}", Arguments.fromBundle(messageObj))
+        Arguments.fromBundle(Utils.convertJsonToBundle(JSONObject(messageStr)))
+      for (view in visitableViews) {
+        view.handleMessage(messageObj)
+      }
     }
   }
 
-  inner class StradaJavaScriptInterface {
-    @JavascriptInterface
-    fun postMessage(messageStr: String) {
-      val messageJSONObj = JSONObject(messageStr)
-      val messageObj = Utils.convertJsonToBundle(messageJSONObj)
-      val eventName = messageJSONObj.getString("component")
-      sendMessage(eventName, Arguments.fromBundle(messageObj))
-    }
+  // region SessionCallbackAdapter
+
+  override fun onReceivedError(errorCode: Int) {
+    visitableViews.last().onReceivedError(errorCode)
+  }
+
+  override fun onRenderProcessGone() {
+    visitableViews.last().onRenderProcessGone()
+  }
+
+  override fun onZoomReset(newScale: Float) {
+    visitableViews.last().onZoomReset(newScale)
+  }
+
+  override fun onZoomed(newScale: Float) {
+    visitableViews.last().onZoomed(newScale)
+  }
+
+  override fun visitCompleted(completedOffline: Boolean) {
+    visitableViews.last().visitCompleted(completedOffline)
   }
+
+  override fun visitLocationStarted(location: String) {
+    visitableViews.last().visitLocationStarted(location)
+  }
+
+  override fun visitProposedToLocation(location: String, options: TurboVisitOptions) {
+    visitableViews.last().visitProposedToLocation(location, options)
+  }
+
+  override fun visitRendered() {
+    visitableViews.last().visitRendered()
+  }
```

Comment:
Could we end up with `visitableViews.last()` returning `null` in the case that a VisitableView disappears and thus is unregistered before a visit is finished? In that case, would we call `visitCompleted()` on a null object?